kyo-direct/shared/src/main/scala/kyo/Validate.scala

```diff
@@ -1,23 +1,114 @@
+                    "dropWhile",
+                    "map",
+                    "flatMap",
+                    "filter",
+                    "filterNot",
+                    "flatten",
+                    "groupBy",
+                    "groupMap",
+                    "fold",
+                    "groupMapReduce",
+                    "count",
+                    "foldRight",
+                    "foldLeft",
+                    "forall",
+                    "foreach",
+                    "exists",
+                    "corresponds",
+                    "find",
+                    "maxByOption",
+                    "scanLeft",
+                    "scanRight",
+                    "span",
+                    "takeWhile"
+                ).contains(methodName)
+
+            validType && validName
+        end validAsyncShift
+
+        extension (inline e: Expr[Any] | Tree)
+            transparent inline def safeShow: String =
+                inline e match
+                    case expr: Expr[Any] => expr.asTerm.safeShow
+                    case tree: Tree =>
+                        if tree.symbol.exists && (tree.symbol.maybeOwner ne Symbol.noSymbol) then
+                            tree.show
+                        else
+                            s"<unnamed:${tree.getClass.getSimpleName.stripSuffix("$")}>"
+        end extension
+
+        def asyncShiftDive(qualifiers: List[Tree])(using Trees.Step): Unit =
+            qualifiers match
+                case List(Block(List(DefDef(_, _, _, Some(body))), _)) =>
+                    body match
+                        case Match(_, cases) =>
+                            cases.foreach:
+                                case CaseDef(_, _, body) => Trees.Step.goto(body)
+                        case _ => Trees.Step.goto(body)
+
+                case _ => qualifiers.foreach(qual => Trees.Step.goto(qual))
+
+        Trees.traverseGoto(expr.asTerm) {
+            case Apply(Apply(TypeApply(select: Select, _), argGroup0), argGroup1) if validAsyncShift(select) =>
+                Trees.Step.goto(select.qualifier)
+                asyncShiftDive(argGroup0)
+                asyncShiftDive(argGroup1)
+
+            case Apply(select: Select, argGroup0) if validAsyncShift(select) =>
+                Trees.Step.goto(select.qualifier)
+                asyncShiftDive(argGroup0)
+
+            case Apply(TypeApply(select: Select, _), argGroup0) if validAsyncShift(select) =>
+                Trees.Step.goto(select.qualifier)
+                asyncShiftDive(argGroup0)
 
-        Trees.traverse(expr.asTerm) {
             case Apply(TypeApply(Ident("now" | "later"), _), List(qual)) =>
+                @tailrec
+                def dive(qual: Tree): Unit =
+                    qual match
+                        case Block(quals, last) =>
+                            quals.foreach(Trees.Step.goto)
+                            dive(last)
+                        case _ =>
```

Comment:
improve current checks to restart traversing the tree after a match that doesn't necessarily lead to a failure.